### PR TITLE
Feature: multipath message & resubmit, conversation placeholder, separate title generation, and some bug fix.

### PR DESCRIPTION
**Pull request by wtlyu**
A basic implementation of resubmit.

It behave not 100% like official, as listed:

1. official will save all messages, this commit will replace the original one and deleted all messages after it.
2. official support multi-path record (like <1/3>), but in this commit it's not implemented yet.
3. official use textarea to work with input, borderless. but in this commit, it use contenteditable, with border.

### CODE REVIEW

**Comment by wtlyu** [comment]
btw, please do a fully test, since it modified a lot.

submission structure has been modified

```

    const submission = {
      isCustomModel,
      message: { 
        sender: 'User',
        text: message, 
        isCreatedByUser: true,
        model,
        chatGptLabel,
        promptPrefix,
      },
      messages,
      currentMsg,
      initialResponse,
      sender,
    };
```

**Comment by wtlyu** [comment]
BTW, another difference

official site allow edit more than one at the same time, but in this commit is not. (not necessary. all msgs will be deleted)

**Comment by danny-avila** [comment]
Thanks, that's a great PR! I had ideas about how to support the multi-path record but maybe we can start with this, and maybe you also have ideas. I'll be testing this out.

File: client/src/components/Messages/Message.jsx [hunk 1]
```diff
@@ -80,6 +94,49 @@ export default function Message({
 
   const wrapText = (text) => <TextWrapper text={text} />;
 
+  const resubmitMessage = () => {
+    const text = textEditor.current.innerHTML
```
Comment (danny-avila) [inline client/src/components/Messages/Message.jsx]: innerHTML adds the HTML tags to the visible text and API response: 
`<div>How do I make modular code in python? what is the convention</div>`

Changing this to `innerText` fixes it

File: client/src/components/Messages/Message.jsx [hunk 2]
```diff
@@ -96,22 +153,47 @@ export default function Message({
...
-                {/* <div className={`${blinker ? 'result-streaming' : ''} markdown prose dark:prose-invert light w-full break-words`}> */}
-                <div className="markdown prose dark:prose-invert light w-full break-words">
-                  {notUser ? wrapText(text) : text}
-                  {blinker && <span className="result-streaming">█</span>}
+            ) : 
+              edit ? (
+                <div className="flex min-h-[20px] flex-col flex-grow items-start gap-4 whitespace-pre-wrap">
+                  {/* <div className={`${blinker ? 'result-streaming' : ''} markdown prose dark:prose-invert light w-full break-words`}> */}
+                  
+                  <div className="markdown prose dark:prose-invert light w-full break-words border-none"
```
Comment (danny-avila) [inline client/src/components/Messages/Message.jsx]: Adding the twcss class 'focus:outline-none' will remove the blue border to emulate the official site.

File: api/server/routes/ask.js [hunk 3]
```diff
@@ -53,6 +53,12 @@ router.post('/', async (req, res) => {
     }
   }
 
+  if (id) {
```
Comment (danny-avila) [inline api/server/routes/ask.js]: Not sure why but conversations are not updating in the Convo NavBar once submitted or when another convo is clicked (expected behavior), whether in a new chat or resubmitted. Maybe id is being kept the same throughout api calls? 

I noticed this in a new chat, and the same is true for resubmissions. This was the start of the convo:
image
I sent another message and then refreshed and noticed it got split into a separate convo:
image

**Comment by danny-avila** [comment]
Do you mind merging master to your branch to include my latest updates, too?

Thank you for the work on this, it's a complex feature and I question how utilized it is by users of the official site. However, in the vein of reverse-engineering their UI, I'd like to have multi-path 'branching' as the goal.

One idea I had to keep track of 'branches' is [populating subdocuments](https://mongoosejs.com/docs/populate.html#populate_multiple_documents). The mongoose schema for messages would need to be updated with a self-referencing relationship to other Messages, i.e.:

```
const messageSchema = new mongoose.Schema({
  ...
  sharedPaths: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Message'
  }]
});
```

Something along those lines.

**Comment by wtlyu** [comment]
> 

looks too complicated. why not do multiatch on frontend？ just started from the first one.

File: client/src/components/Messages/Message.jsx (hunk 1 above)
Comment (wtlyu) [reply client/src/components/Messages/Message.jsx]: fixed

File: api/server/routes/ask.js (hunk 3 above)
Comment (wtlyu) [reply api/server/routes/ask.js]: fixed

**Comment by wtlyu** [comment]
Hi @danny-avila , after asking for more friends that I find out some bugs.

1. in short, I agree that we should make some multi-path work, deleting and replace message will lead to some corner case.
2. by the way, I remove the id from message, because messageId can be decided by us, so don't need it anymore.
3. I change the ask.js to achieve: return userMessage with random messageID and conversationID before calling the ChatGpt API, return both userMessage and gptResponse at the end, chaining parentMessageId to always exist, parentMessageId must be chain from the lastest message to the first message, the first level message should have parentMessageId = 0000-00000000-0000... 
4. BTW, I move the title generation into another separate call, in order to speed up.
5. and now we will have a conversation placeholder, even though the first response get error. (conversationId is generated by server, not API server anymore)
6. for now, the resubmit will cause a new message to the end of the list, but I am working on frontend part to show something like <1/3>
7. fix some small bugs, code changes too lot, I'm lazy to create new PR for small fixes.

**Comment by wtlyu** [comment]
ONE BIG PROBLEM

I donnot have api to test bing version and browser version, only chatgpt and customGpt is tested. 

I change the code of bing version and browser version at same time, but we need someone to test.

**Comment by wtlyu** [comment]
TODO

- [ ]  test on bing and browser version
- [ ] frontend part of multi-path messages
- [ ] error message if we go back to a conversation. (userMessage will be saved once /api/ask, not after response finished, so we may face a userMessage without responseMessage)

**Comment by wtlyu** [comment]
> Do you mind merging master to your branch to include my latest updates, too?
> 
> Thank you for the work on this, it's a complex feature and I question how utilized it is by users of the official site. However, in the vein of reverse-engineering their UI, I'd like to have multi-path 'branching' as the goal.
> 
> One idea I had to keep track of 'branches' is [populating subdocuments](https://mongoosejs.com/docs/populate.html#populate_multiple_documents). The mongoose schema for messages would need to be updated with a self-referencing relationship to other Messages, i.e.:
> 
> ```
> const messageSchema = new mongoose.Schema({
>   ...
>   sharedPaths: [{
>     type: mongoose.Schema.Types.ObjectId,
>     ref: 'Message'
>   }],
>   descendants: [{
>     type: mongoose.Schema.Types.ObjectId,
>     ref: 'Message'
>   }]
> });
> ```
> 
> Something along those lines (and of course the frontend handling would be pretty involved

unnecessary to be honest.
we have messageID and parentMessageId in out data structure, and I have tested that:
1. we can assign messageId of userMessage, in request
2. we can assign conversationId in request
3. within same conversationId, if same messageId request, the conversation will be ignored and create a new conversation ---> so replace the old message and reset it, using same id will create a new conversation.
4. not sure what will happen if the pre-assigned conversationId conflict, but it's mostly not happen
5. parentMessageId can be assigned by us, if parentMessageId is null, new conversation will be created, conversationId will be ignored. ---> So I create new conversation in db, generate conversationId, then send the first message, with parentMessageId = 0000-00000000-0000.... that's ok, and no check for meaning less id.

So that, if we deal with parentMessageId correctly, we can use it to chain all messages. and the top-level messages（maybe one or more）, will have parentMessageId = 0000-00000000-0000. Everything else is frontend part.

**Comment by wtlyu** [comment]
I think the multi-path messages are ready for testing. Including lots of small fix.

multi-path messages, including resubmit and new message:
QQ20230313-225602

I also make the conversation placeholder works, and move title generation to a separate call.
QQ20230313-230135

**Comment by wtlyu** [comment]
> > Do you mind merging master to your branch to include my latest updates, too?
> > Thank you for the work on this, it's a complex feature and I question how utilized it is by users of the official site. However, in the vein of reverse-engineering their UI, I'd like to have multi-path 'branching' as the goal.
> > One idea I had to keep track of 'branches' is [populating subdocuments](https://mongoosejs.com/docs/populate.html#populate_multiple_documents). The mongoose schema for messages would need to be updated with a self-referencing relationship to other Messages, i.e.:
> > ```
> > const messageSchema = new mongoose.Schema({
> >   ...
> >   sharedPaths: [{
> >     type: mongoose.Schema.Types.ObjectId,
> >     ref: 'Message'
> >   }],
> >   descendants: [{
> >     type: mongoose.Schema.Types.ObjectId,
> >     ref: 'Message'
> >   }]
> > });
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Something along those lines (and of course the frontend handling would be pretty involved
> 
> unnecessary to be honest. we have messageID and parentMessageId in out data structure, and I have tested that:
> 
> 1. we can assign messageId of userMessage, in request
> 2. we can assign conversationId in request
> 3. within same conversationId, if same messageId request, the conversation will be ignored and create a new conversation ---> so replace the old message and reset it, using same id will create a new conversation.
> 4. not sure what will happen if the pre-assigned conversationId conflict, but it's mostly not happen
> 5. parentMessageId can be assigned by us, if parentMessageId is null, new conversation will be created, conversationId will be ignored. ---> So I create new conversation in db, generate conversationId, then send the first message, with parentMessageId = 0000-00000000-0000.... that's ok, and no check for meaning less id.
> 
> So that, if we deal with parentMessageId correctly, we can use it to chain all messages. and the top-level messages（maybe one or more）, will have parentMessageId = 0000-00000000-0000. Everything else is frontend part.

Here is how my implementation work:

1. change the server side to ensure that, every message will have a messageId: userMessage is generated by server, and responseMessage is generated by gpt server
2. every message should have the parentMessageId, responseMessage will target the userMessage, userMessage will target the previous responseMessage, toplevel message will target 0000-00000000-0000....
3. even though new message in frontend, which maybe have no real messageId and parentMessageId yet. The submit function will generate a fake messageId to the placeholder message.
4. in Messages/index.js, once messages update, will build a messageTree based on messageId and parentMessageId, the structure is like: 

```
[ // root level, more then one top level message,
    { messageId: 'example1', parentMessageId: '0000-00000000-0000...(too long)', children: [
        { messageId: 'example2', parentMessageId: 'example1', children: []},
        { messageId: 'example3', parentMessageId: 'example1', children: []},
    ]},
    { messageId: 'example11', parentMessageId: '0000-00000000-0000...(too long)', children: [
        { messageId: 'example22', parentMessageId: 'example11', children: [
            { messageId: 'example222', parentMessageId: 'example22', children: []},
            { messageId: 'example333', parentMessageId: 'example22', children: []},
        ]},
        { messageId: 'example33', parentMessageId: 'example11', children: []},
    ]},
]
```

5. Messages will render a MultiMessage component, who holds the siblingIdx and control which child should be rendered. (siblingIdx actually is reverse number, siblingIdx=0 means the last one. that is the default path)
6. if a Message component with message.children == [], that means this is the last one, and it's messageId will be saved in Conversation.parentMessageId, that will be used as parent to send new message.
7. If a message is a placeholder message and the gpt is currently generating it, it will have a message.submitting = true flag. So blink will relay on this.

BTW, the position of sibling switch is conflict with the username "User:", so I replace the username using a similar icon like chatgpt and bing.

image

image

image

**Comment by danny-avila** [comment]
Thanks for your hard work! I'm currently at work but I will review and test your changes as soon as I can

File: api/models/Message.js [hunk 4]
```diff
@@ -1,7 +1,7 @@
 const mongoose = require('mongoose');
 
 const messageSchema = mongoose.Schema({
-  id: {
+  messageId: {
```
Comment (danny-avila) [inline api/models/Message.js:4]: This is a breaking change. I understand it but prior messages are now unavailable

**Comment by danny-avila** [comment]
gpt responses are not being saved: 
image

creates empty conversation
```
MongoServerError: Plan executor error during findAndModify :: caused by :: E11000 duplicate key error collection: chatgpt-client.messages index: id_1 dup key: { id: null }
    at Connection.onMessage (C:\Users\Daniela\Documents\Personal\Coding\clone-forks\chatgpt-clone\api\node_modules\mongodb\lib\cmap\connection.js:230:30)
    at MessageStream.<anonymous> (C:\Users\Daniela\Documents\Personal\Coding\clone-forks\chatgpt-clone\api\node_modules\mongodb\lib\cmap\connection.js:61:60)
    at MessageStream.emit (node:events:513:28)
    at processIncomingData (C:\Users\Daniela\Documents\Personal\Coding\clone-forks\chatgpt-clone\api\node_modules\mongodb\lib\cmap\message_stream.js:125:16)
    at MessageStream._write (C:\Users\Daniela\Documents\Personal\Coding\clone-forks\chatgpt-clone\api\node_modules\mongodb\lib\cmap\message_stream.js:33:9)
    at writeOrBuffer (node:internal/streams/writable:392:12)
    at _write (node:internal/streams/writable:333:10)
    at Writable.write (node:internal/streams/writable:337:10)
    at Socket.ondata (node:internal/streams/readable:766:22)
    at Socket.emit (node:events:513:28) {
  ok: 0,
  code: 11000,
  codeName: 'DuplicateKey',
  keyPattern: { id: 1 },
  keyValue: { id: null },
  [Symbol(errorLabels)]: Set(0) {}
}
```

File: api/models/Message.js (hunk 4 above)
Comment (wtlyu) [reply api/models/Message.js:4]: yes. I'm new to mongoose. do they have some migration script? like django orm.

**Comment by wtlyu** [comment]
> 

How to reproduce this?

**Comment by wtlyu** [comment]
> 

will refer to the error message, it seems a null of message.id is created. It's what you said a break change.

Do you have any idea on this? At lease delete database and rebuild will fix

**Comment by danny-avila** [comment]
> > 
> 
> will refer to the error message, it seems a null of message.id is created. It's what you said a break change.
> 
> Do you have any idea on this? At lease delete database and rebuild will fix

since I am using the same db as used with the latest master branch, it is determining omission of id as a null value and so a duplicate every time there is no id provided

**Comment by wtlyu** [comment]
yep. I might face the same error, but after rebuild the db, it disappeared.

So does mongoose have something like migrations? for example, we can delete the original messageId field, and then rename the id field to messageId field. So that the old data can also be reserved.

**Comment by danny-avila** [comment]
> yep. I might face the same error, but after rebuild the db, it disappeared.
> 
> So does mongoose have something like migrations? for example, we can delete the original messageId field, and then rename the id field to messageId field. So that the old data can also be reserved.

I think this would fix it, setting a default value, and you can have it generate a new UUID 
https://stackoverflow.com/questions/7617002/dealing-with-schema-changes-in-mongoose

**Comment by wtlyu** [comment]
> > yep. I might face the same error, but after rebuild the db, it disappeared.
> > So does mongoose have something like migrations? for example, we can delete the original messageId field, and then rename the id field to messageId field. So that the old data can also be reserved.
> 
> I think this would fix it, setting a default value, and you can have it generate a new UUID https://stackoverflow.com/questions/7617002/dealing-with-schema-changes-in-mongoose

you mean set default on id? or messageId?

**Comment by danny-avila** [comment]
> > > yep. I might face the same error, but after rebuild the db, it disappeared.
> > > So does mongoose have something like migrations? for example, we can delete the original messageId field, and then rename the id field to messageId field. So that the old data can also be reserved.
> > 
> > 
> > I think this would fix it, setting a default value, and you can have it generate a new UUID https://stackoverflow.com/questions/7617002/dealing-with-schema-changes-in-mongoose
> 
> you mean set default on id? or messageId?

default on messageId

**Comment by wtlyu** [comment]
MARK, a possible bug on this PR to fix

If a submitting message is failed or cancelled, then click regenerate or resubmit the request, will create a new conversation.

**Comment by wtlyu** [comment]
> > > > yep. I might face the same error, but after rebuild the db, it disappeared.
> > > > So does mongoose have something like migrations? for example, we can delete the original messageId field, and then rename the id field to messageId field. So that the old data can also be reserved.
> > > 
> > > 
> > > I think this would fix it, setting a default value, and you can have it generate a new UUID https://stackoverflow.com/questions/7617002/dealing-with-schema-changes-in-mongoose
> > 
> > 
> > you mean set default on id? or messageId?
> 
> default on messageId

but this can not fix your error message. it says a id=1 conflict with id=null. I think it is because I delete the id field, but previous this field is created and marked as unique.

**Comment by danny-avila** [comment]
> > > > > yep. I might face the same error, but after rebuild the db, it disappeared.
> > > > > So does mongoose have something like migrations? for example, we can delete the original messageId field, and then rename the id field to messageId field. So that the old data can also be reserved.
> > > > 
> > > > 
> > > > I think this would fix it, setting a default value, and you can have it generate a new UUID https://stackoverflow.com/questions/7617002/dealing-with-schema-changes-in-mongoose
> > > 
> > > 
> > > you mean set default on id? or messageId?
> > 
> > 
> > default on messageId
> 
> but this can not fix your error message. it says a id=1 conflict with id=null. I think it is because I delete the id field, but previous this field is created and marked as unique.

You're right, there's no straightforward migration sync for schema changes in mongodb. I wrote my own script that I will push to your branch

**Comment by wtlyu** [comment]
> MARK, a possible bug on this PR to fix
> 
> If a submitting message is failed or cancelled, then click regenerate or resubmit the request, will create a new conversation.

fixed.

as side effect, now database will save the error message, and display error message by anytime.

**Comment by danny-avila** [comment]
there are a couple more bugs, also migration related, with respect to conversation. also for some reason the modelmenu is experiencing an error with remembering the last selected model (could be related again to migrating issue). Do you mind adding me as a collaborator on your fork so I can help with some of these?

**Comment by danny-avila** [comment]
I'm finish the migration script but wanted to bring something up to your attention. I uploaded this on #50 too.

Your method for the placeholder works but if a new chat is started before it's finished, it never gets titled; this is reflected in the DB too : 

```
{
...,
  "parentMessageId": "00000000-0000-0000-0000-000000000000",
  "suggestions": [],
  "title": "New Chat",
}
```
master branch:

https://user-images.githubusercontent.com/110412045/224863041-6dc73c40-6a43-4892-bee3-9aa7cf9e546f.mp4

fork:

https://user-images.githubusercontent.com/110412045/224863061-c4a44faa-cdf3-4ed8-b15f-b389b0a9d090.mp4

**Comment by danny-avila** [comment]
You can see some of my edits to your code here: https://github.com/danny-avila/chatgpt-clone/tree/fix-bugs

The migration code is in models/Conversation.js. Tested and working, but will test it on a copy of my larger db soon.

Also Sydney/Bing is not working at all on your fork, couldn't get to it today.

In the future, we could do smaller PRs and more incrementally to troubleshoot better. I've yet to make a test suite and regretting it now haha

**Comment by wtlyu** [comment]
> there are a couple more bugs, also migration related, with respect to conversation. also for some reason the modelmenu is experiencing an error with remembering the last selected model (could be related again to migrating issue). Do you mind adding me as a collaborator on your fork so I can help with some of these?

sent

**Comment by wtlyu** [comment]
> 

something related to the official version

In the official version, it behave like this:
1. If you open the conversation untill it finished, then a call /gen_title will be called, and new title changed.-> requested by frontend?
2. If you leave the conversation before finished, you will see a request is still in background, but the /gen_title will never called. But after refresh, the conversation has the new title -> requested by backend
3. If you leave the conversation, then go back before finished, You will see the userMessage but no response anymore, shows "There was an error generating a response". But actually ,refersh will get the right response.
4. If you refresh the conversation before finish, the request is cancelled, you might get a error message in this conversation, or get a not finished result. (BUT how? how they get the mesageId, if the response is not finished)

So this should be ok

1. in ask.js called res.end() once response finished, (or gen title cost time)
2. then ask.js call the gen_title in backend.
3. For the frontend, once a new response received, and it's the first level message, (not strictly restricted the first one). will call /gen_title.
4. there are options on /gen_title implementation: only return the database title or make a duplicate title generation. Since title generation may take seconds, I suggest the duplicate. (or frontend will not received the right result)

**Comment by wtlyu** [comment]
> 

fixed

**Comment by danny-avila** [comment]
Great, have you pushed those commits? I'm actively working on your branch and would like to merge your changes. Yes, reorganizing the ask endpoints is overdue, there is a lot of code repetition I was actually in the process of refactoring haha

**Comment by danny-avila** [comment]
I really like your multi-path implementation btw. I will probably refactor to organize the logic when I get the chance, for example, the messageTree can be created in the redux reducer for messages, but I look forward to merging this

**Comment by wtlyu** [comment]
> 

Thank you, I have some unknown trouble pushing to github, working on fix...

**Comment by wtlyu** [comment]
I think my work is almost done. the remaining thing is to do more test and fix them.
@danny-avila will you testing this? It's too 4am for me, I have to sleep

**Comment by danny-avila** [comment]
Hey @wtlyu ,

I didn't mean to push my changes to your branch, sorry about that. See my commits for what I worked on. 

On model change, conversation will reset. This has an interesting side effect where you can cycle through the last convo in the current convo. This wasn't intended and could be removed with messing with the messageTree; however I kinda like it! It's only temporary and on refresh or convo click, it goes away

Here's a video showing this: https://i.imgur.com/Pw9PT2O.mp4

There were a lot of errors with custom models. I fixed as many as I could, only one remaining
image

The custom model is set to user icon. After we fix this, I think we're ready to merge

**Comment by danny-avila** [comment]
It's a little bugger on second thought, so i think we could mess with the message tree to get it showing right without being able to cycle to last conversation

**Comment by wtlyu** [comment]
Update: 

1. fix the bug that gptCustom icon shows as User
2. add icon into model menu
3. add tooltip to icon
4. fix a bug that clean the convo, even though model not changed (click the same one)
5. now change model will auto clear all messages. in case for unexpected behavior 


icon in model menu
image

tooltip of icon
image


also fix #68 #64

**Comment by wtlyu** [comment]
I think I have nothing on my todo list. I will turn to rebase feat-usersys to match this branch.

**Comment by wtlyu** [comment]
> using the latest state of your branch, the icons are not rendering as squares anymore in the chat view, but I can work on this.
> 
> Great work. there is a lot of refactoring to do since a lot of the code is looking messy (I admit a lot of it was messy beforehand), also unused variables in a lot of the code and many things that can go in their own file. I will remove my eslint file from gitignore so we can following linting rules going forward. These are not big issues now but can make it difficult to navigate the code and debug

I'm sharing your feelings. It's somehow confusing code to me, especially how to strictly distinguish which part belongs to message or conversation.

**Comment by danny-avila** [comment]
closing #55

**Pull request merged**